kernel/sched.c

```diff
@@ -449,19 +454,17 @@ void z_sched_start(struct k_thread *thread)
  * another CPU to catch the IPI we sent and halt.  Note that we check
  * for ourselves being asynchronously halted first to prevent simple
  * deadlocks (but not complex ones involving cycles of 3+ threads!).
+ * Acts to release the provided lock before returning.
  */
-static k_spinlock_key_t thread_halt_spin(struct k_thread *thread, k_spinlock_key_t key)
+static void thread_halt_spin(struct k_thread *thread, k_spinlock_key_t key)
 {
 	if (is_halting(_current)) {
 		halt_thread(_current,
 			    is_aborting(_current) ? _THREAD_DEAD : _THREAD_SUSPENDED);
 	}
-	k_spin_unlock(&_sched_spinlock, key);
 	while (is_halting(thread)) {
+		arch_spin_relax();
 	}
-	key = k_spin_lock(&_sched_spinlock);
-	z_sched_switch_spin(thread);
-	return key;
 }
 
```

Comment:
The new version of thread_halt_spin() is unfortunately wrong as sched_spinlock is no longer being manipulated by this routine. Not unlocking (or releasing) it before the `while (is_halting(thread))` loop prevents another CPU from performing the abort/suspend the target thread (and also leaves the `key` argument unused).

Reply:
Oh, for... yeah, I pushed a bad tree.  Looks like I dropped a git add or something and that one line is still sitting in my tree uncommited.  Also it looks like I had an incorrect cut/paste when redoing the commit message of the main patch and it ended up being a copy of the test patch instead.  One sec...